### ISSUE TITLE
Some trailing comma coverage still missing

### ISSUE THREAD

**Issue by jidanni** [perltidy/perltidy#143]
I wanted a trailing comma:
```
$ cat z.pl
my %strips = (
    1 => [
        [ [ 1750, 150, ], [ 1850, 150, ], ],
        [ [ 1950, 150, ], [ 2050, 150, ], ],
    ] # <---------- here
);
```
Alas, no matter what I tried, I am not getting it,
```
$ for i in  0 1 m b h i; do perltidy --noprofile --add-trailing-commas \
  --standard-output --want-trailing-commas=$i z.pl|diff - z.pl; done
$
```

**Comment by shancock9** [comment]
The comma which would be added would be the only comma at its level. There is a fixed rule that 
the ``--add-trailing-commas`` parameter can not add a new comma in that case.

On the other hand, if there had been a comma, the ``--delete-trailing-commas`` parameter could
delete it. But if it did, it could not be replaced, as mentioned in the manual for the example

```
   my ( $self, ) = @_;
```

and

```
   my ( $self ) = @_;
```

The same rule applies if the term ``$self`` is replaced an item with any lower level complexity.

**Comment by dboehmer** [comment]
> There is a fixed rule that the `--add-trailing-commas` parameter can not add a new comma in that case.

Just thinking: Could this rule be improved either
1. by detecting hash assignments like `%hash = ( one_key => 42 )`, `%{ $hashref } = ( one_key => 42 )` or `$hashref->%* = ( one_key => 42 )` or
2. by treating a list with 2 elements and the array operator `( key => "value" )` differently?

**Comment by shancock9** [comment]
Yes, I've been looking into option 2, including fat commas in the comma count. If it works I'll post a tagged version.

**Comment by shancock9** [comment]
Tagged version 20240511.02 has an update to fix this

https://github.com/perltidy/perltidy/tags

**Comment by jidanni** [comment]
Perhaps somewhere on the man page it could be mentioned that a big motivation for adding trailing commas is to be able to add more code later easier.

**Comment by Julien-Elie** [comment]
Regarding trailing comma coverage, would it also be possible to add the trailing comma with `-atc -wtc=m` in code like:

```
    $self->{thread} = App::DocKnot::Spin::Thread->new(
        {
            output      => $output,
            sitemap     => $self->{sitemap},
            source      => $input,
            'style-url' => $self->{style_url},
            versions    => $self->{versions},
        }  # <---------- here
    );
```

**Comment by shancock9** [comment]
It would be difficult to add, but I'll look at it.

**Comment by shancock9** [comment]
An issue with adding this comma is that it will prevent the ``--weld-nested-containers``
flag from working. So it requires a new flag to be added to be sure that this is what
the user really wants. The new flag is ``--add-lone-trailing-comma, -altc``. 

This is in tagged Tagged version 20240511.06 at https://github.com/perltidy/perltidy/tags.

So adding -altc gives

```
# perltidy -atc -altc -wtc=m
self->{thread} = App::DocKnot::Spin::Thread->new(
    {
        output      => $output,
        sitemap     => $self->{sitemap},
        source      => $input,
        'style-url' => $self->{style_url},
        versions    => $self->{versions},
    },
);

```

**Comment by Julien-Elie** [comment]
Just tested.  Works fine, many thanks!
The improved related documentation is also great.

**Comment by Julien-Elie** [comment]
Thanks for the latest 20240903 release Steve!  The 100th one in Github 👏🏻 

FWIW, here are other missing trailing commas I've recently found out.  Just sharing them with you in case you could take them into account.

```
    logmsg(
        sprintf(
            "Article %s: processed notice %s by %s for %s "
              . "(%d ids, %d skipped, %.5f s, %.1f/s)",
            $msgid, $nid, $issuer, $type, $nr, $skipped, $diff,
            ($nr + $skipped) / $diff,
        ) # <---------- here
    );
```

```
    my ($set_cutofflow_in, $set_cutofflow_out) = argparser(
        [
            {
                name    => "cutofflow",
                default => 0,
            } # <---------- here
        ],
        [
            {
                name => "errmsg",
            } # <---------- here
        ], 
    );
```

```
    if (
        system(
            "$INN::Config::newsbin/shlock", '-p', $$, '-f', $lockfile # <---------- here
        ) == 0
    ) {
        $lockfiles{$lockfile} = 1;
        return 1;
    }
```

```
    parse_config(
        {
            '<anywhere>' => {
                'identd'          => 'string',
                'password'        => 'string',
                'patterns'        => 'string',
                'hostname'        => 'list of hostnames',
                'hold-time'       => 'number',
                'max-connections' => 'number / "unlimited" / "none"',
                'ignore'          => 'boolean',
                'list'            => 'boolean',
                'resendid'        => 'boolean',
                'skip'            => 'boolean',
                'streaming'       => 'boolean',
                'xbatch'          => 'boolean',
            },
            'group' => {},
            'peer'  => {},
        } # <---------- here
    );
```

```
    $getgroupid = $dbh->prepare(
        q{
select groupid from groupinfo
    where cast(groupname as text) = ?1
        and deleted = 0;
} # <---------- here
    );
```

**Comment by shancock9** [comment]
Thanks for all the help in testing.  I'll see if some of these cases can be included in the trailing comma logic.
The main issue is to verify that instability cannot occur, regardless of user input parameters.

**Comment by Julien-Elie** [comment]
Thanks Steve for this update.  Amongst the 5 last samples I provided, it handles the 1st, 2nd and 4th.
👍🏻

**Comment by shancock9** [comment]
Thanks for testing this. The 3rd example is not currently possible because it is near a stability limit.
The 5th is not possible because it is the end of a multiline q quote and gets processed specially.

**Comment by Julien-Elie** [comment]
Re-running the CodeLayout::RequireTrailingCommaAtNewline perlcritic policy, I found out another case:

```
my %groups = parse_config(
    {
        'method' => {
            'class'      => 'number',
            'size'       => 'minsize[,maxsize] definition',
        } # <---------- here
    },
);
```

Incidentally, if I manually add the comma and then run `perltidy -wtc=m -altc -atc -dtc`, the comma remains.  Isn't it supposed to be removed as it is not added if not present?

I see it is the same behaviour for the 3rd and 5th examples: if I manually add the colon, it remains.  Is it normal?  (I tested with version 20240903.02.)

**Comment by shancock9** [comment]
It should work with ``--wtc=b`` (bare comma) instead of ``wtc=m``. The ``m`` option still needs
some clarification of its definition for these types of commas which require the ``altc`` flag.  

It is normal for existing commas to be kept in cases like these involving involving
a single list item, even if they cannot be added. This is because adding commas is a riskier
operation than keeping commas, from a stability standpoint. So the requirements for adding
a comma are more strict than for keeping one.

**Comment by Julien-Elie** [comment]
OK, I understand.
`--wtc=b` did not remove the comma.  I had to use `--wtc=0`.
Interestingly, I then re-ran `--wtc=m` and also tried `--wtc=h` but some parts of the code did not appear as before (the ones with a single hash on one line).

```
my ($set_cutofflow_in, $set_cutofflow_out) = argparser(
    [
        {
            name    => "cutofflow",
            default => 0,
        },
    ],
    [
        {
            name => "errmsg",
        }
    ],
);
```

is now reformatted this way, and the comma in the latest hash is not re-added, even with `-altc -atc`:

```
my ($set_cutofflow_in, $set_cutofflow_out) = argparser(
    [
        {
            name    => "cutofflow",
            default => 0,
        },
    ],
    [{ name => "errmsg" }],
);
```

I thought `--wtc=h` would add them?  (the bare trailing commas)

**Comment by shancock9** [comment]
The flag ``-wtc=b`` means "want bare trailing commas", so ``-wtc=b -dtc`` would
not remove a bare trailing comma. An alternative to using ``-wtc=0 -dtc``,
which removes all trailing commas, the ``-dwic`` flag could be used to remove
just the special lone trailing commas (also called weld interfering commas),
and leave the rest alone.

The flags ``-wtc=i`` and ``wtc=h`` only apply to bare trailing commas (they are
special cases of ``-wtc=b``), so they would not add a comma inside the braces
in this line

```
    [{ name => "errmsg" }],
```

Thanks for the feedback.  I am working on some improvements and expect to
release a new version in a couple of weeks, along with some expanded
documentation which tries to clarify some of these points.

**Comment by Julien-Elie** [comment]
Thanks for the improvements in the new 20240903.03 release.  It handles the 3rd case discussed above 😄 
The new details in documentation are great and extensive.  Thanks for them!

**Comment by Julien-Elie** [comment]
Yes indeed, as far as I am concerned, I am totally happy with the latest improvements regarding the handling of trailing commas.  FWIW, I am using [the following options](https://github.com/InterNetNews/inn/blob/main/.perltidyrc):

```
-drc            # delete repeated commas in a list
-wtc=m          # add trailing commas to all multi-line lists
-atc            # add trailing commas where needed
-dtc            # delete trailing commas where needed
-altc           # add trailing comma to long arguments (perlcritic needs this)
-dwic           # delete commas that interfere with welding
```

These new options also made the maintainer of the podlators package happy: https://github.com/rra/podlators/commit/002a6bafb7a1266f7406a26df3e041361031af0c.  The result is slightly better to the eyes than what perlcritic enforces.

Thanks again Steve for all your work.
The new `-wnt` flag is also great; it helped to [detect and improve](https://github.com/InterNetNews/inn/commit/ecc9940da67ca1d2024d6ffa38ff8a0380111ebd) the formatting of a nested ternary expression.

**Comment by shancock9** [comment]
Thanks for the note! And it's nice to see that the new ``--warn-nested-ternaries`` is useful.
I added it one day when I realized I was wasting too much time reading some of my own ternaries,
and I needed a way to track them down to clean them up.
